app/controllers/api/submission_files_controller.rb

```diff
@@ -199,10 +163,91 @@ def remove_folder
       end
     end
 
+    def submit_file
+      student = Student.find_by(user: @real_user, course: @current_course)
+      assignment = Assignment.find_by(id: params[:assignment_id], course: @current_course)
+
+      # Disable submission via API if the instructor desires to
+      unless assignment.api_submit
+        render 'shared/http_status', locals: { code: '403', message:
+          'The instructor has disabled submission via the API' }, status: :forbidden
+        return
+      end
+
+      # Reject submission if the student has pending grouping
+      if student.has_pending_groupings_for?(assignment.id)
+        render 'shared/http_status', locals: { code: '422', message:
+          'You must respond to your group request on MarkUs before you can submit' }, status: :unprocessable_entity
+        return
+      end
+
+      grouping = if student.has_accepted_grouping_for?(assignment.id)
+                   student.accepted_grouping_for(assignment.id)
+                 elsif assignment.group_max == 1
+                   student.create_group_for_working_alone_student(assignment.id)
+                   student.accepted_grouping_for(assignment.id)
+                 else
+                   student.create_autogenerated_name_group(assignment)
+                 end
+
+      upload_file(grouping, only_required_files: assignment.only_required_files)
+    end
+
     protected
 
     def implicit_authorization_target
       SubmissionFile
     end
+
+    private
+
+    # Helper that uploads a file to this particular +grouping+'s assignment repository.
+    # If +only_required_files+ is true, only required files for this grouping's assignment can be uploaded.
+    def upload_file(grouping, only_required_files: false)
+      if has_missing_params?([:filename, :mime_type, :file_content])
```

Comment:
I know the Jupyter Extension uses the `upload` parameter and I have tested it using the parameter. However, to recycle more code and to keep with the format we have for `create` I thought it might be a good idea to keep these parameters and modify the parameters the extension sends instead.

Let me know though if you disagree with this. I can change it easily either way.